extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesTaskRunner.java

```diff
@@ -0,0 +1,571 @@
+  {
+    client.cleanUpJob(new K8sTaskId(taskid));
+  }
+
+
+  @Override
+  public Optional<InputStream> streamTaskReports(String taskid)
+  {
+    return Optional.absent();
+  }
+
+  @Override
+  public List<Pair<Task, ListenableFuture<TaskStatus>>> restore()
+  {
+    return ImmutableList.of();
+  }
+
+  @Override
+  public void start()
+  {
+    log.info("Starting cleanup executor for jobs older than 1 day....");
+    cleanupExecutor.scheduleAtFixedRate(
+        () ->
+            client.cleanCompletedJobsOlderThan(
+                KubernetesTaskRunnerConfig.toMilliseconds(k8sConfig.taskCleanupDelay),
+                TimeUnit.MILLISECONDS
+            ),
+        1,
+        KubernetesTaskRunnerConfig.toMilliseconds(k8sConfig.taskCleanupInterval),
+        TimeUnit.MILLISECONDS
+    );
+    log.info("Started cleanup executor for jobs older than 1 day....");
+  }
+
+
+  @Override
+  public void stop()
+  {
+    log.info("Stopping...");
+    cleanupExecutor.shutdownNow();
+    log.info("Stopped...");
+  }
+
+  @Override
+  public Map<String, Long> getTotalTaskSlotCount()
+  {
+    return ImmutableMap.of("taskQueue", (long) taskQueueConfig.getMaxSize());
+  }
+
+  private List<String> javaOpts(Task task)
+  {
+    final List<String> javaOpts = new ArrayList<>();
+    Iterables.addAll(javaOpts, k8sConfig.javaOptsArray);
+
+    // Override task specific javaOpts
+    Object taskJavaOpts = task.getContextValue(
+        ForkingTaskRunnerConfig.JAVA_OPTS_PROPERTY
+    );
+    if (taskJavaOpts != null) {
+      Iterables.addAll(
+          javaOpts,
+          new QuotableWhiteSpaceSplitter((String) taskJavaOpts)
+      );
+    }
+
+    javaOpts.add(StringUtils.format("-Ddruid.port=%d", DruidK8sConstants.PORT));
+    javaOpts.add(StringUtils.format("-Ddruid.plaintextPort=%d", DruidK8sConstants.PORT));
+    javaOpts.add(StringUtils.format("-Ddruid.tlsPort=%d", node.isEnableTlsPort() ? DruidK8sConstants.TLS_PORT : -1));
+    javaOpts.add(StringUtils.format(
+        "-Ddruid.task.executor.tlsPort=%d",
+        node.isEnableTlsPort() ? DruidK8sConstants.TLS_PORT : -1
+    ));
+    javaOpts.add(StringUtils.format("-Ddruid.task.executor.enableTlsPort=%s", node.isEnableTlsPort())
+    );
+    return javaOpts;
+  }
+
+  private List<String> generateCommand(Task task)
+  {
+    final List<String> command = new ArrayList<>();
+    // must use deep storage shuffle now with k8s peon tasks.
+
+    command.add("/peon.sh");
+    command.add(taskConfig.getTaskDir(task.getId()).toString());
+    command.add("1"); // the attemptId is always 1, we never run the task twice on the same pod.
+
+    String nodeType = task.getNodeType();
+    if (nodeType != null) {
+      command.add("--nodeType");
+      command.add(nodeType);
+    }
+
+    // If the task type is queryable, we need to load broadcast segments on the peon, used for
+    // join queries
+    if (task.supportsQueries()) {
+      command.add("--loadBroadcastSegments");
+      command.add("true");
```

Comment:
General question: if a peon supports queries, it will have incoming connections. How do we ensure that the Peon's IP address is visible to the Broker? K8s usually has some kind of overlay network. Are we assuming Druid runs within K8s so it is on that same network? Or, that somewhere we've set up a proxy into the overlay network? It's been a while since I used K8s, perhaps the proxy isn't needed? Or, the networks are merged in AWS? Is that true on plain-Jane K8s?

Reply:
Good question.  For this patch, I am forcing the tasks to run in the current namespace that druid is running in.  You could easily have the tasks run in a different namespace but would require a bit of changes to this patch (not much though).  So when a job gets launched, we wait for the pod to startup, once the pod starts, we grab the ip address of the pod and broadcast that as the `TaskLocation`.  There would be no need for a proxy if you wanted to go across namespaces in the same context, but if you had druid running in namespace `A` and tasks running in namespace `B` you would have to setup your k8s to be able to communicate bidirectionally.

Reply:
The goal of this patch, is if you have are running druid on `k8s`, you shouldn't need to do anything except add the extension and a few configuration options to the overlord and you are good to go.  In the gotchas section of the readme, there are some things you have to look out for with sidecars which i mentioned.   Also your service account must have the ability to deal with k8s jobs, I included a sample role / rolebinding as an example in the readme.md.  We have been using this on some of our clusters for a while now and we haven't really had to add any extra configuration or do anything special in k8s to get this to work.